### PR TITLE
feat: normalized nft slice part 2

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Within this scope:

- [x] Consolidate NFT item and collection data so that we don't need (but keep it as a convenience join / intermediary value for parsers) a `WithCollection` flavor anymore - ensure we properly update collection in case of *missing* fields only a la `updateItems`
- [x] ensure we still fetch collection meta as a separate query from the NFT modal, not from the NFT tab, which should use the collection metadata coming from the items resolution, the same as currently

Out of scope

- Selectors join with the NFTs from assetsSlice - and by extension, using selectors (other than nftApi `.select()`) - it seems like we're sometimes getting different contract addresses from Alchemy/Zapper and unchained e.g  https://etherscan.io/address/0x57f1887a8bf19b14fc0df6fd9b2acc9af147ea85 from Zapper/Alchemy (ENS contract) and https://etherscan.io/address/0xc18360217d8f7ab5e7c516566761ea12ce7f9d72  from Unchained/Etherscan (registar)
- Consolidates the initial (effectively only, something something) RTK query dispatching for `getNftUserTokens` in `<AppContext />` - the component-level firing is safer for now. We may introduce the unification into a follow-up draft, to ensure we're not spamming services, though from a high-level overfetching perspective, it probably makes sense to be lazy vs. eager here and I believe `<AppContext />` may actually *not* be a good place for this

Consuming `selectNftsWithCollection` instead of the RTK query breaks the notion of NFTs by AccountId, thus switching wallets is borked if we consume it and would keep previous NFTs, a.k.a something something two hard things in software engineering.
We DO need the join from portfolio if we want to use selectors (which we do) be able to still discriminate by held NFTs (which we also obviously do).
The consumption of the join cannot be done for the time being and will be for a follow-up (a.k.a normalization part 3)

tl;dr we still need to consume `nftApi` vs *akschual*. nft selectors for the time being :sadpepe:

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Testing

- Regression test NFT tab/avatar selection, and ensure everything looks good against prod

### Engineering

- Normalization looks good

### Operations

- ☝🏽

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4563** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4563" target="_blank">Graphite</a> 
    * **PR #4574** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4574" target="_blank">Graphite</a> 
      * **PR #4582** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4582" target="_blank">Graphite</a> 
        * **PR #4575** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4575" target="_blank">Graphite</a> 
          * **PR #4596** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4596" target="_blank">Graphite</a> 
            * **PR #4597** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4597" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4597?utm_source=stack-comment).

File: src/state/apis/nft/types.ts [hunk 1]
```diff
@@ -32,7 +32,11 @@ export type NftItem = {
   description: string
   // TODO(gomes): store the collection data into nft[collectionAssetId] and make this `collectionId`
   // we can then join the item and collection, once the collection is in the store as well
-  collection: NftCollectionType
+  collectionId: AssetId
   medias: MediaUrl[]
   rarityRank: number | null
 }
+
+export type NftItemWithCollection = Omit<NftItem, 'collectionId'> & {
```
Comment (gomesalexandre) [inline src/state/apis/nft/types.ts:40]: Temporary type, revert before opening me

File: src/components/Modals/Nfts/NftModal.tsx [hunk 2]
```diff
@@ -77,24 +78,32 @@ export const NftModal: React.FC<NftModalProps> = ({ nftItem }) => {
...
-  const { data: nftCollection } = useGetNftCollectionQuery(
-    { accountIds, collectionId: nftItem.collection.id ?? '' },
-    { skip: !nftItem.collection.id },
+  const nftCollectionParams = useMemo(
+    () => ({ assetId: nftItem.collectionId }),
+    [nftItem.collectionId],
   )
 
+  // TODO(gomes): check if we still need to fire the RTK query
+  // const { data: nftCollection } =
```
Comment (gomesalexandre) [inline src/components/Modals/Nfts/NftModal.tsx:87]: We do, revert me

File: src/state/apis/nft/types.ts (hunk 1 above)
Comment (gomesalexandre) [reply src/state/apis/nft/types.ts:40]: Or perhaps not, so far it turns out to be useful as an intermediary value

File: src/components/Nfts/NftCard.tsx [hunk 3]
```diff
@@ -19,48 +19,59 @@ import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingl
...
 }
 
-export const NftCard: React.FC<NftCardProps> = ({ nftItem }) => {
-  const { collection, medias, name, rarityRank } = nftItem
-  const { floorPrice } = collection
+export const NftCard: React.FC<NftCardProps> = ({ nftAssetId }) => {
+  const nftFilter = useMemo(() => ({ assetId: nftAssetId }), [nftAssetId])
+  const nftItem = useAppSelector(state => selectNftById(state, nftFilter))
+
+  if (!nftItem) throw new Error(`NFT not found for assetId: ${nftAssetId}`)
```
Comment (gomesalexandre) [inline src/components/Nfts/NftCard.tsx:35]: This should never happen but it may

File: src/state/apis/nft/types.ts (hunk 1 above)
Comment (gomesalexandre) [reply src/state/apis/nft/types.ts:39]: Keeping this as an intermediary type for parsers, but also as a return type for `selectNftItemsWithCollection`, to be used as a join selector for components to consume, so this can play nicely with `matchSorter`

File: src/components/Nfts/NftCard.tsx [hunk 4]
```diff
@@ -19,48 +19,59 @@ import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingl
...
+
+  const { collectionId, medias, name, rarityRank } = nftItem
+
+  const collection = useAppSelector(state =>
+    selectNftCollectionById(state, { assetId: collectionId }),
```
Comment (gomesalexandre) [inline src/components/Nfts/NftCard.tsx:40]: memo filter

File: src/components/Modals/Nfts/components/NftOverview.tsx [hunk 5]
```diff
@@ -10,23 +10,24 @@ import { Row } from 'components/Row/Row'
...
 
-export const NftOverview: React.FC<NftOverviewProps> = ({ nftItem, nftCollection }) => {
+export const NftOverview: React.FC<NftOverviewProps> = ({ nftItem }) => {
   const translate = useTranslate()
 
-  const description =
-    nftItem.description || nftItem.collection.description || nftCollection?.description
-  const collection = nftItem?.collection
+  const collection = useAppSelector(state =>
+    selectNftCollectionById(state, { assetId: nftItem?.collectionId }),
```
Comment (gomesalexandre) [inline src/components/Modals/Nfts/components/NftOverview.tsx:26]: memo filter

File: src/state/apis/nft/nftApi.ts [hunk 6]
```diff
@@ -116,23 +129,42 @@ export const nftApi = createApi({
...
+
+        const nftsById = data.reduce<NftState['nfts']['byId']>((acc, item) => {
+          if (!item.collection.assetId) return acc
+
+          const nftAssetId: AssetId = `${item.collection.assetId}/${item.id}`
+          let { collection, ...nftItemWithoutId } = item
+          const nftItem: NftItem = {
+            ...nftItemWithoutId,
+            collectionId: item.collection.assetId,
+          }
```
Comment (gomesalexandre) [inline src/state/apis/nft/nftApi.ts:142]: from denormalized to normalized

File: src/state/apis/nft/selectors.ts [hunk 7]
```diff
@@ -3,21 +3,59 @@ import { createSelector } from 'reselect'
...
+import type { NftItemWithCollection } from './types'
+
+const selectNfts = (state: ReduxState) => state.nft.nfts.byId
+const selectNftCollections = (state: ReduxState) => state.nft.collections.byId
+
+export const selectNftItemsWithCollection = createSelector(
+  selectNfts,
+  selectNftCollections,
+  (nfts, collections): NftItemWithCollection[] => {
+    return Object.values(nfts).reduce<NftItemWithCollection[]>((acc, nft) => {
```
Comment (gomesalexandre) [inline src/state/apis/nft/selectors.ts:15]: Join on NFTs and collections data, we can't use portfolio for this yet even though that's what we should really do, a join of portfolio NFT to NFT to collection, see PR description

File: src/components/Modals/Nfts/NftModal.tsx [hunk 8]
```diff
@@ -283,18 +284,14 @@ export const NftModal: React.FC<NftModalProps> = ({ nftItem }) => {
...
             <TabPanel p={0}>
               <NftCollection
-                name={nftCollection?.name || nftItem.collection.name}
-                description={nftCollection?.description || nftItem.collection.description}
-                socialLinks={
-                  nftCollection?.socialLinks.length
-                    ? nftCollection?.socialLinks
-                    : nftItem.collection.socialLinks
-                }
+                name={nftCollection?.name || nftCollection?.name}
```
Comment (kaladinlight) [inline src/components/Modals/Nfts/NftModal.tsx:292]:
```suggestion
                name={nftCollection?.name}
```

File: src/components/Modals/Nfts/NftModal.tsx [hunk 9]
```diff
@@ -263,13 +267,10 @@ export const NftModal: React.FC<NftModalProps> = ({ nftItem }) => {
...
   const nftModalDetails = useMemo(() => {
-    if (!(nftCollection || nftItem.collection)) return null
+    if (!nftCollection) return null
 
     const hasUsefulCollectionData = Boolean(
-      nftItem.collection.description ||
-        nftCollection?.description ||
-        nftItem.collection.socialLinks?.length ||
-        nftCollection?.socialLinks?.length,
+      nftCollection?.description || nftCollection?.socialLinks?.length,
```
[the remark this answers is not in the document]
Comment (kaladinlight) [reply src/components/Modals/Nfts/NftModal.tsx:273]: shouldn't need the `?` optional chain as well with the check on line 270?

File: src/components/Modals/Nfts/NftModal.tsx [hunk 10]
```diff
@@ -312,7 +309,7 @@ export const NftModal: React.FC<NftModalProps> = ({ nftItem }) => {
     )
   }, [nftModalOverview, nftModalDetails])
 
-  if (!(nftCollection || nftItem.collection)) return null
+  if (!(nftCollection || nftItem.collectionId)) return null
```
Comment (kaladinlight) [inline src/components/Modals/Nfts/NftModal.tsx:312]: seems like an unnecessary check

Comment (gomesalexandre) [reply src/components/Modals/Nfts/NftModal.tsx]: https://github.com/shapeshift/web/pull/4597/commits/9a49919b80b3ea62a4b8518a4162ce49a48966ab

File: src/state/apis/covalent/covalentApi.ts [hunk 11]
```diff
@@ -34,52 +34,54 @@ export const covalentApi = createApi({
...
 
-        // Covalent is used only for Polygon NFTs for now
-        const chainId = polygonChainId
-        const network = chainIdToCovalentNetwork(chainId)
-        let data: NftItem[] = []
-        const limit = 100
+          // Covalent is used only for Polygon NFTs for now
+          const chainId = polygonChainId
+          const network = chainIdToCovalentNetwork(chainId)
+          let data: NftItemWithCollection[] = []
```
Comment (kaladinlight) [inline src/state/apis/covalent/covalentApi.ts:47]: nitpick: can use `const` I think if you use `push`

File: src/components/Modals/Nfts/NftModal.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/Modals/Nfts/NftModal.tsx]: Will add the name in a follow-up since this would change the flow - agreed on opt-chain 9459632397270be80be0bc5d3c55a12440b8c980

File: src/components/Modals/Nfts/NftModal.tsx (hunk 8 above)
Comment (gomesalexandre) [reply src/components/Modals/Nfts/NftModal.tsx]: derp https://github.com/shapeshift/web/pull/4597/commits/f99be919045304b26c94824b23f2bf769e6f23a6

File: src/components/Nfts/NftCard.tsx [hunk 12]
```diff
@@ -19,48 +19,56 @@ import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingl
...
+  const floorPrice = collection?.floorPrice
   const mediaUrl = medias?.[0]?.originalUrl
   const mediaType = getMediaType(mediaUrl)
   const bg = useColorModeValue('gray.50', 'gray.750')
   const bgHover = useColorModeValue('gray.100', 'gray.700')
   const placeholderImage = useColorModeValue(PlaceholderDrk, Placeholder)
   const [isMediaLoaded, setIsMediaLoaded] = useState(false)
 
-  const chainId = collection.chainId
+  const chainId = collection?.chainId
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/Nfts/NftCard.tsx]: c6407c0db

File: src/components/Modals/Nfts/NftModal.tsx [hunk 13]
```diff
@@ -283,18 +284,14 @@ export const NftModal: React.FC<NftModalProps> = ({ nftItem }) => {
...
               <NftCollection
-                name={nftCollection?.name || nftItem.collection.name}
-                description={nftCollection?.description || nftItem.collection.description}
-                socialLinks={
-                  nftCollection?.socialLinks.length
-                    ? nftCollection?.socialLinks
-                    : nftItem.collection.socialLinks
-                }
+                name={nftCollection.name}
+                description={nftCollection?.description}
```
Comment (kaladinlight) [inline src/components/Modals/Nfts/NftModal.tsx:293]: one more optional chain that can be removed

Comment (gomesalexandre) [reply src/components/Modals/Nfts/NftModal.tsx]: 7c0b1b6df

File: src/components/Nfts/NftCard.tsx [hunk 14]
```diff
@@ -19,48 +19,56 @@ import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingl
...
     const eventData = {
       name: nftItem.name,
       id: nftItem.id,
-      collectionName: nftItem.collection.name,
-      collectionAddress: fromAssetId(nftItem.collection.assetId!).assetReference,
+      collectionName: collection?.name,
+      collectionAddress: fromAssetId(collection?.assetId).assetReference,
       price: nftItem.price,
-      collectionFloorPrice: nftItem.collection.floorPrice,
+      collectionFloorPrice: collection?.floorPrice,
```
Comment (kaladinlight) [inline src/components/Nfts/NftCard.tsx:66]: unnecessary optional chaining on collection if we want to clean up

File: src/state/apis/nft/selectors.ts [hunk 15]
```diff
@@ -3,24 +3,66 @@ import { createSelector } from 'reselect'
...
+      return acc
+    }, [])
+  },
+)
+
+export const selectNftCollectionById = createSelector(
+  selectNftCollections,
+  (_state: ReduxState, assetId: AssetId) => assetId,
+  (collections, collectionAssetId): NftCollectionType | undefined => {
+    if (!collectionAssetId) return undefined
```
Comment (kaladinlight) [inline src/state/apis/nft/selectors.ts:34]:
```suggestion
    if (!collectionAssetId) return
```

File: src/state/apis/nft/selectors.ts [hunk 16]
```diff
@@ -3,24 +3,66 @@ import { createSelector } from 'reselect'
...
-
 export const selectSelectedNftAvatarUrl = createSelector(
   selectSelectedNftAvatar,
   selectNfts,
-  (nftAssetId, nfts): string | null => {
-    const nftItem = nfts.byId[nftAssetId ?? '']
+  (nftAssetId, nfts): string | undefined => {
+    const nftItem = nfts[nftAssetId ?? '']
 
     const media: string | undefined = nftItem?.medias[0]?.originalUrl
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/state/apis/nft/selectors.ts]: f46a4940b7426a32bf0150e2f4ad38dbe40cd0a8

File: src/state/apis/nft/selectors.ts (hunk 15 above)
Comment (gomesalexandre) [reply src/state/apis/nft/selectors.ts]: 774089500a3436adb5f55166f41230bb7b5215b8

**Review by kaladinlight — APPROVED** [review]
